packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +95,235 @@ export function matches(element: Element, selector: AST): boolean {
+  // sanity check, since the distance is provided as part of filter
+  if (distance <= 0 || distance >= 256) {
+    return null;
+  }
+  let ancestor: Element | null = element;
+  for (let i = 0; i < distance; i++) {
+    ancestor = ancestor.parentElement;
+    if (ancestor === null) {
+      return null;
+    }
+  }
+  return ancestor;
+}
 
-  if (
-    selector.type === 'id' ||
-    selector.type === 'class' ||
-    selector.type === 'type' ||
-    selector.type === 'attribute'
-  ) {
-    elements.push(...element.querySelectorAll(selector.content));
-  } else if (selector.type === 'list') {
-    for (const subSelector of selector.list) {
-      elements.push(...querySelectorAll(element, subSelector));
+function findAncestorBySelector(element: Element, selector: string): Element | null {
+  let ancestor: Element | null = element.parentElement;
+  while (ancestor !== null) {
+    if (ancestor.matches(selector)) {
+      return ancestor;
     }
-  } else if (selector.type === 'compound') {
-    // TODO - handling compound needs to be reworked...
-    // .cls:upward(1) for example will not work with this implementation.
-    // :upward is not about selecting, but transforming a set of nodes (i.e.
-    // uBO's transpose method).
-    if (selector.compound.length !== 0) {
-      elements.push(
-        ...querySelectorAll(element, selector.compound[0]).filter((e) =>
-          selector.compound.slice(1).every((s) => matches(e, s)),
-        ),
+    ancestor = ancestor.parentElement;
+  }
+  return null;
+}
+
+function handleCompoundSelector(element: Element, compound: AST[]): Element[] {
+  if (compound.length === 0) {
+    return [];
+  }
+
+  const upwardIndex = compound.findIndex((s) => s.type === 'pseudo-class' && s.name === 'upward');
+
+  if (upwardIndex === -1) {
+    const [firstSelector, ...restSelectors] = compound;
+    return querySelectorAll(element, firstSelector).filter((e) =>
+      restSelectors.every((s) => matches(e, s)),
+    );
+  }
+  const before = compound.slice(0, upwardIndex);
+  const upward = compound[upwardIndex];
+  const after = compound.slice(upwardIndex + 1);
+
+  const candidates =
+    before.length > 0
+      ? querySelectorAll(element, { type: 'compound', compound: before })
+      : [element];
+
+  const ancestors = new Set<Element>();
+
+  for (const candidate of candidates) {
+    if (upward.type !== 'pseudo-class' || upward.name !== 'upward') {
+      continue;
+    }
+    const { argument } = upward;
+    if (argument === undefined) {
+      continue;
+    }
+
+    const distance = parseInt(argument, 10);
+    const ancestor = !Number.isNaN(distance)
+      ? findAncestorByDistance(candidate, distance)
+      : findAncestorBySelector(candidate, argument);
+
+    if (ancestor === null) {
+      continue;
+    }
+    ancestors.add(ancestor);
+  }
+
+  if (ancestors.size === 0) {
+    return [];
+  }
+
+  if (after.length > 0) {
```

Comment:
Could we here fallback to `handleCompoundSelector` instead of `querySelectorAll`? In pseudo code:

```
result_set = empty
for each ancestor
  result_set += handleCompoundSelector(ancestor, after);
```